File: REPASO PRUEBA U3/RepasoPrueba/src/main/java/ec/edu/espe/repasoprueba/controller/PaintingController.java
package ec.edu.espe.repasoprueba.controller;

import ec.edu.espe.repasoprueba.model.Painting;
import ec.edu.espe.repasoprueba.utils.MongoConnection;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.Filters;
import com.mongodb.client.model.Updates;
import com.mongodb.client.result.DeleteResult;
import com.mongodb.client.result.UpdateResult;
import org.bson.Document;
import java.util.ArrayList;
import java.util.List;

/**
 *
 * @author Mathews Pastor, The POOwer Rangers Of Programming
 */
public class PaintingController {

    private MongoCollection<Document> collection;
    private static final double IVA_RATE = 0.15;

    public PaintingController() {
        // Obtenemos la conexión y la colección específica
        MongoDatabase db = MongoConnection.getInstance().getDatabase();
        this.collection = db.getCollection("paintings");
    }

    public double calculateIva(double price) {
        double result = price * (1 + IVA_RATE);
        return Math.round(result * 100.0) / 100.0; // Redondeo a 2 decimales
    }

    // --- CREATE ---
    public boolean createPainting(String id, String name, double price, List<String> colors) {
        try {
            double priceWithIva = calculateIva(price);

            Document doc = new Document("id", id)
                    .append("name", name)
                    .append("price", price)
                    .append("colors", colors)
                    .append("priceWithIva", priceWithIva);

            collection.insertOne(doc);
            return true;
        } catch (Exception e) {
            System.err.println("Error creating painting: " + e.getMessage());
            return false;
        }
    }

    // --- READ ---
    public List<Painting> getAllPaintings() {
        List<Painting> list = new ArrayList<>();
        for (Document doc : collection.find()) {
            try {
                list.add(mapDocumentToPainting(doc));
            } catch (Exception e) {
                System.err.println("Error: " + e.getMessage());
            }
        }
        return list;
    }

// --- FIND ---
    public Painting findPaintingById(String id) {
        Document doc = collection.find(Filters.eq("id", id)).first();
        if (doc != null) {
            return mapDocumentToPainting(doc);
        }
        return null;
    }

    // --- UPDATE ---
    public boolean updatePainting(String id, String name, double price, List<String> colors) {
        try {
            double priceWithIva = calculateIva(price);

            // Ejecutamos la actualización
            UpdateResult result = collection.updateOne(
                    Filters.eq("id", id), 
                    Updates.combine(
                            Updates.set("name", name),
                            Updates.set("price", price),
                            Updates.set("colors", colors),
                            Updates.set("priceWithIva", priceWithIva)
                    )
            );
            // Verificamos "getMatchedCount". Si es 0, es que NO encontró el ID,
            // por lo tanto no actualizó nada.
            return result.getMatchedCount() > 0;

        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
    }
    
    
// --- DELETE ---
    public boolean deletePainting(String id) {
        try {
            DeleteResult result = collection.deleteOne(Filters.eq("id", id));
            // Si getDeletedCount es 0, significa que el ID no existía en la BD.
            return result.getDeletedCount() > 0;

        } catch (Exception e) {
            System.err.println("Error deleting: " + e.getMessage());
            return false;
        }
    }

    // --- MAPPER ---
    private Painting mapDocumentToPainting(Document doc) {
        //Id que pone el Mongo
        Object idObj = doc.get("id");
        String id;

        if (idObj != null) {
            id = idObj.toString(); // Convertimos lo que sea a String
        } else {
            // Fallback: Si no tiene campo "id", usamos el "_id" de Mongo
            id = doc.getObjectId("_id").toString();
        }

        String name = doc.getString("name");
        Number priceNum = (Number) doc.get("price");
        double price = (priceNum != null) ? priceNum.doubleValue() : 0.0;

        List<String> colors = doc.getList("colors", String.class);

        Number ivaNum = (Number) doc.get("priceWithIva");
        Double storedIva = (ivaNum != null) ? ivaNum.doubleValue() : null;

        if (storedIva == null) {
            storedIva = calculateIva(price);
        }

        return new Painting(id, name, price, colors, storedIva);
    }
}

File: REPASO PRUEBA U3/PY/controller/painting_controller.py
from utils.mongo_connection import MongoConnection
from model.painting import Painting

class PaintingController:
    def __init__(self):
        self.db = MongoConnection.get_database()
        self.collection = self.db["paintings"]
        self.IVA_RATE = 0.15

    def calculate_iva(self, price):
        result = price * (1 + self.IVA_RATE)
        return round(result, 2)

    # --- CREATE ---
    def create_painting(self, id_painting, name, price, colors):
        try:
            price_with_iva = self.calculate_iva(price)
            
            doc = {
                "id": id_painting,
                "name": name,
                "price": price,
                "colors": colors,
                "priceWithIva": price_with_iva
            }
            
            self.collection.insert_one(doc)
            return True
        except Exception as e:
            print(f"Error creating painting: {e}")
            return False

    # --- READ ---
    def get_all_paintings(self):
        paintings = []
        try:
            for doc in self.collection.find():
                paintings.append(self._map_document_to_painting(doc))
        except Exception as e:
            print(f"Error reading paintings: {e}")
        return paintings

    # --- FIND ---
    def find_painting_by_id(self, id_painting):
        try:
            doc = self.collection.find_one({"id": id_painting})
            if doc:
                return self._map_document_to_painting(doc)
        except Exception as e:
            print(f"Error finding painting: {e}")
        return None

    # --- UPDATE (Con lógica corregida: matched_count) ---
    def update_painting(self, id_painting, name, price, colors):
        try:
            price_with_iva = self.calculate_iva(price)

            result = self.collection.update_one(
                {"id": id_painting},
                {"$set": {
                    "name": name,
                    "price": price,
                    "colors": colors,
                    "priceWithIva": price_with_iva
                }}
            )
            # Retorna True solo si encontró el documento
            return result.matched_count > 0
        except Exception as e:
            print(f"Error updating: {e}")
            return False

    # --- DELETE (Con lógica corregida: deleted_count) ---
    def delete_painting(self, id_painting):
        try:
            result = self.collection.delete_one({"id": id_painting})
            return result.deleted_count > 0
        except Exception as e:
            print(f"Error deleting: {e}")
            return False

    # --- Helper Mapper ---
    def _map_document_to_painting(self, doc):
        id_val = doc.get("id")
        if id_val is None:
            id_val = str(doc.get("_id"))
        
        name = doc.get("name", "")
        price = doc.get("price", 0.0)
        colors = doc.get("colors", [])
        
        price_iva = doc.get("priceWithIva")
        if price_iva is None:
            price_iva = self.calculate_iva(price)
            
        return Painting(id_val, name, price, colors, price_iva)